### PR TITLE
PyTorch Frontend

### PR DESCRIPTION
**Pull request by tbennun**
PyTorch (2.0 and above) frontend for LBANN. Contains APIs to ingest existing PyTorch modules and functions, and convert them to LBANN graphs, including network parameters.

- [x] `lbann.evaluate` API as a testing and ease-of-use infrastructure
- [x] Add ability to save/load `.protobin` files (especially useful for large models with pretrained parameters)
    - [x] protobin<->prototext converter in `python -m lbann.proto.text2bin` and `bin2text`
- [x] `lbann.torch` module with `compile`/`lazy_compile` API based on PyTorch Dynamo/Inductor
    - [x] Module PyTorch->LBANN replacements
    - [x] Functional PyTorch->LBANN replacements
    - [x] Weight/Parameter conversion procedures
- [x] Tests: 
    - [x] Simple functions and `nn.Module`s
    - [x] Lowering test
    - [x] LeNet-5
    - [x] CosmoFlow
    - [x] ResNet from TIMM
    - [x] ~BERT from HuggingFace Transformers~ HuggingFace unsupported until minibatch dimension can be modified
- [x] Add option to disable Bessel's correction in Batch Normalization to better match PyTorch (only activated by default in the PyTorch frontend)

### CODE REVIEW

File: python/lbann/torch/compiler.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+try:
+    import torch
+    if int(torch.__version__.split('.')[0]) < 2:
+        raise ImportError(
+            'PyTorch 2.0 or newer is required for the LBANN torch frontend. '
+            f'Found version {torch.__version__}')
+except (ModuleNotFoundError, ImportError):
+    raise ImportError(
+        'PyTorch (version 2.0 or newer) is required for the LBANN torch '
+        'frontend.')
```
Comment (ndryden) [inline python/lbann/torch/compiler.py:41]: Does this actually ensure we have Dynamo? I thought you had to have a nightly build or build from source, even with 2.0, to get it.

**Review by ndryden — APPROVED** [review]
This looks really good. I have only really minor comments.

(I haven't tested it myself, but we should land this and then iterate.)

File: python/lbann/torch/compiler.py (hunk 1 above)
Comment (tbennun) [reply python/lbann/torch/compiler.py:41]: It exists when pip installing it. Not sure otherwise.

File: python/lbann/torch/lowering.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+import torch._inductor.compile_fx as inductor
+from torch.fx.passes.fake_tensor_prop import FakeTensorProp
+from torch._subclasses.fake_tensor import FakeTensorMode
+from typing import Any, Dict, List, Optional
+import warnings
+
+
+def dynamo_callback(input_names: List[str], with_weights: bool,
+                    graph_module: fx.GraphModule,
+                    example_inputs: List[torch.Tensor]) -> helpers.LBANNGraph:
```
[the remark this answers is not in the document]
Comment (tbennun) [reply python/lbann/torch/lowering.py:46]: There was no "raises" type hint so this is the best hint I can give to a user on the output type.

**Review by benson31 — APPROVED** [review]
I only reviewed the non-Python portions. Seems fine.

File: ci_test/pytorch_tests/test_lenet.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    # Run experiment
+    lbann.contrib.launcher.run(trainer,
+                               model,
+                               data_reader,
+                               opt,
+                               job_name='lenet_pytorch_test')
+
+
+if __name__ == '__main__':
+    test_lenet_eval()
```
Comment (bvanessen) [inline ci_test/pytorch_tests/test_lenet.py:127]:
```suggestion
    test_lenet_eval()
    test_lenet_train()
```

**Review by bvanessen — APPROVED** [review]
Looks good.  I would just ask that you update the ReleaseNotes document.

**Comment by tbennun** [comment]
@bvanessen Done, please verify release note comment.
